docs: remove TODOs from Static vs Normal Queries to focus more on internals

gillkyle opened the pull request:
## Description

This doc had several TODO's hanging around, there was redundant information around the differences in usage between this doc and the using static query components doc, so I removed it from this section and tried to refocus it more on Gatsby  internals.

I answered what I felt I could discern from reading code in the query-watcher, query-runner, and query-compilation files, but if someone from the Core team has more context and can help realign something it would be very appreciated 😬 

## Related Issues

Closes #15157

muescha commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -10,27 +10,15 @@ title: Static vs Normal Queries
-Started here because they're referenced in page-query-runner:findIdsWithDataDependencies.
-
-The redux `staticQueryComponents` is a map from component jsonName to StaticQueryObject. E.g
+The redux `staticQueryComponents` is a `Map` from component jsonName to StaticQueryObject. An example entry in that data structure looks like this:
```
```suggestion
The redux `staticQueryComponents` is a `Map` from component `jsonName` to `StaticQueryObject`. An example entry in that data structure looks like this:
```

muescha commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -10,27 +10,15 @@ title: Static vs Normal Queries
+## The `staticQueryComponents` and `components` redux stores
 
-## useStaticQuery hook
+Gatsby stores the queries from your site in redux stores called `components` and `staticQueryComponents`. This process and a flowchart illustrating it are explained in the [query extraction](/docs/query-extraction/#store-queries-in-redux) guide.
```
```suggestion
Gatsby stores the queries from your site in Redux stores called `components` and `staticQueryComponents`. This process and a flowchart illustrating it are explained in the [query extraction](/docs/query-extraction/#store-queries-in-redux) guide.
```

marcysutton commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -10,27 +10,15 @@ title: Static vs Normal Queries
 > You can help by making a PR to [update this documentation](https://github.com/gatsbyjs/gatsby/issues/14228).
 
-## How StaticQuery differs from page query
+Static queries differ from the normal Gatsby page queries in a number of ways. One example is that Gatsby is capable of handling queries with variables for pages because of it's awareness of pages while generating them. Doing the same for queries inside of specific components is not yet possible, which is where static queries are used. That means data fetched won't be dynamic (hence the name "sta ...
```
This is a confusing paragraph. Maybe it would be more clear with some examples?
```suggestion
Static queries differ from Gatsby page queries in a number of ways. One example is that Gatsby is capable of handling queries with variables for pages because of its awareness of page context while generating them. Doing the same for queries inside of specific components is not yet possible, which is where static queries are used. That means data fetched won't be dynamic (hence the name "static" query).
```

marcysutton commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -10,27 +10,15 @@ title: Static vs Normal Queries
+Static queries differ from the normal Gatsby page queries in a number of ways. One example is that Gatsby is capable of handling queries with variables for pages because of it's awareness of pages while generating them. Doing the same for queries inside of specific components is not yet possible, which is where static queries are used. That means data fetched won't be dynamic (hence the name "sta ...
 
-StaticQuery can do most of the things that page query can, including fragments. The main differences are:
+_For more information on the practical differences in usage between static and normal queries, refer to the guide on [static query](/docs/static-query/#how-staticquery-differs-from-page-query), this guide discusses the differences in how they are handled internally by Gatsby_
```
```suggestion
_For more information on the practical differences in usage between static and normal queries, refer to the guide on [static query](/docs/static-query/#how-staticquery-differs-from-page-query). This guide discusses the implementation differences in how they are handled internally by Gatsby_
```

marcysutton commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -10,27 +10,15 @@ title: Static vs Normal Queries
-- StaticQuery does not work with raw React.createElement calls; please use JSX, e.g. `<StaticQuery />`
-  - _NOTE: you can also use the new `useStaticQuery` hook; more information below_
-- Static Queries don't need to get run for each page.(ie:Just once)
+## The `staticQueryComponents` and `components` redux stores
```
```suggestion
## The `staticQueryComponents` and `components` Redux stores
```

marcysutton commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -10,27 +10,15 @@ title: Static vs Normal Queries
-Started here because they're referenced in page-query-runner:findIdsWithDataDependencies.
-
-The redux `staticQueryComponents` is a map from component jsonName to StaticQueryObject. E.g
+The redux `staticQueryComponents` is a `Map` from component `jsonName` to `StaticQueryObject`. An example entry in that data structure looks like this:
```
```suggestion
In Redux, `staticQueryComponents` is a `Map` from component `jsonName` to `StaticQueryObject`. An example entry in that data structure looks like this:
```

marcysutton commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -45,18 +33,12 @@ The redux `staticQueryComponents` is a map from component jsonName to StaticQuer
+In the example above, `blog-2018-07-17-announcing-gatsby-preview-995` is the key, with the object as its value in the `Map`.
 
-### Usages
+The `staticQueryComponents` redux namespace is owned by the `static-query-components.js` reducer which reacts to `REPLACE_STATIC_QUERY` actions. It is created in `query-watcher.js` and is called in [`websocket-manager.js`](https://github.com/gatsbyjs/gatsby/blob/610b5812a815f9ecff422e9087c851cd103c8e7e/packages/gatsby/src/utils/websocket-manager.js#L85) to watch for updates to querys while you ar ...
```
```suggestion
The `staticQueryComponents` Redux namespace is owned by the `static-query-components.js` reducer which reacts to `REPLACE_STATIC_QUERY` actions. It is created in `query-watcher.js` and called in [`websocket-manager.js`](https://github.com/gatsbyjs/gatsby/blob/610b5812a815f9ecff422e9087c851cd103c8e7e/packages/gatsby/src/utils/websocket-manager.js#L85) to watch for updates to queries while you are developing, and add new data to your cache when queries change.
```

marcysutton commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -45,18 +33,12 @@ The redux `staticQueryComponents` is a map from component jsonName to StaticQuer
 
-- [websocket-manager](#TODO). TODO
-- [query-watcher](#TODO).
+## Other related internal methods used static queries
```
```suggestion
## Other related internal methods using static queries
```
Does that work better?

marcysutton commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -45,18 +33,12 @@ The redux `staticQueryComponents` is a map from component jsonName to StaticQuer
+The `getQueriesSnapshot` function returns a `Map` with a snapshot of both the `staticQueryComponents` and `components` that holds all queries from Redux to watch.
 
-  TODO: Finish above
+The `handleComponentsWithRemovedQueries` function removes static queries from Redux when they no longer have a query associated with them, which is derived from any staticQueryComponent which doesn't have a `componentPath` on it.
```
This is a pretty awkward sentence, and I'm not sure how to improve it.
```suggestion
The `handleComponentsWithRemovedQueries` function removes static queries from Redux when they no longer have a query associated with them. This derives from any staticQueryComponent which doesn't have a `componentPath` on it.
```

gillkyle commented:
Just updated the PR to tell more of a story of the code like Marcy suggested, as well as removed a couple lines that me and @sidharthachatterjee didn't think really provide much new context or insight.

Added some more code snippets to showcase the idea of what's changing in your code being built as well.

laurieontech commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -2,35 +2,21 @@
+## Differences between varieties of GraphQL queries
 
-StaticQuery can do most of the things that page query can, including fragments. The main differences are:
+Static queries differ from Gatsby page queries in a number of ways. Gatsby is capable of handling queries with variables for pages because of its awareness of page context while generating them, and **page queries can only be made in top-level page components**.
```
```suggestion
Static queries differ from Gatsby page queries in a number of ways. For pages, Gatsby is capable of handling queries with variables because of its awareness of page context. However, **page queries can only be made in top-level page components**.
```

laurieontech replied:
Overall, this is really helpful context for understanding the differences.
Feel free to disagree with the edits, it just felt a bit more readable this way.
I also wonder what we can do to better define "page context"?

laurieontech commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -2,35 +2,21 @@
-- StaticQuery does not work with raw React.createElement calls; please use JSX, e.g. `<StaticQuery />`
-  - _NOTE: you can also use the new `useStaticQuery` hook; more information below_
-- Static Queries don't need to get run for each page.(ie:Just once)
+Doing the same for queries inside of specific components lower in the component tree is not yet possible, which is where static queries are used. Data fetched with a static query won't be dynamic (i.e. **they can't use variables**, hence the name "static" query), but they can be called at any level in the component tree.
```
```suggestion
In contrast, static queries do not take variables. This is because static queries are used inside specific components, and can appear lower in the component tree. Data fetched with a static query won't be dynamic (i.e. **they can't use variables**, hence the name "static" query), but they can be called at any level in the component tree.
```

laurieontech commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -45,18 +31,97 @@ The redux `staticQueryComponents` is a map from component jsonName to StaticQuer
+
+## Replacing queries with JSON imports
+
+With the final build, there isn't a GraphQL server running to query for data. Gatsby has already [extracted](/docs/query-extraction/) and [run](/docs/query-execution/) the queries, [storing](/docs/query-execution/#save-query-results-to-redux-and-disk) them in files based on hashes in `/public/static/d/<hash>.json`. It can now remove code for GraphQL queries, because there isn't a server it can qu ...
```
```suggestion
With the final build, there isn't a GraphQL server running to query for data. Gatsby has already [extracted](/docs/query-extraction/) and [run](/docs/query-execution/) the queries, [storing](/docs/query-execution/#save-query-results-to-redux-and-disk) them in files based on hashes in `/public/static/d/<hash>.json`. It can now remove code for GraphQL queries, because the data is already available.
```

laurieontech commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -45,18 +31,97 @@ The redux `staticQueryComponents` is a map from component jsonName to StaticQuer
+### Distinguishing between static and normal queries
 
-TODO: in query-watcher.js/handleQuery, we remove jsonName from dataDependencies. How did it get there? Why is jsonName used here, but for other dependencies, it's a path?
+Babel traverses all of your source code looking for queries during query extraction, in order for Gatsby to handle static and normal queries differently, it looks for 3 specific cases in [`babel-plugin-remove-graphql-queries`](https://github.com/gatsbyjs/gatsby/blob/master/packages/babel-plugin-remove-graphql-queries/src/index.js):
```
```suggestion
Babel traverses all of your source code looking for queries during query extraction. In order for Gatsby to handle static and normal queries differently, it looks for 3 specific cases in [`babel-plugin-remove-graphql-queries`](https://github.com/gatsbyjs/gatsby/blob/master/packages/babel-plugin-remove-graphql-queries/src/index.js):
```

laurieontech commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -45,18 +31,97 @@ The redux `staticQueryComponents` is a map from component jsonName to StaticQuer
 
-  - `getQueriesSnapshot` returns map with snapshot of `state.staticQueryComponents`
-  - handleComponentsWithRemovedQueries. For each staticQueryComponent, if passed in queries doesn't include `staticQueryComponent.componentPath`. TODO: Where is StaticQueryComponent created? TODO: Where is queries passed into `handleComponentsWithRemovedQueries`?
+Code specific for querying with GraphQL is no longer relevant and can be swapped out.
```
Is a word missing here? I'm not sure what this is saying.

gillkyle replied:
Yeah this is a little terse, here's a suggestion:

```suggestion
Code that is specific for querying the GraphQL server set up during build time is no longer relevant, and can be swapped out in exchange for the JSON data that has been extracted for each query.
```

laurieontech commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -45,18 +31,97 @@ The redux `staticQueryComponents` is a map from component jsonName to StaticQuer
-  - handleComponentsWithRemovedQueries. For each staticQueryComponent, if passed in queries doesn't include `staticQueryComponent.componentPath`. TODO: Where is StaticQueryComponent created? TODO: Where is queries passed into `handleComponentsWithRemovedQueries`?
+Code specific for querying with GraphQL is no longer relevant and can be swapped out.
+
+A static query written in a component with the `useStaticQuery` hook like this:
```
Not sure what the transition here is? Maybe a brief introduction about this being an example?

gillkyle replied:
```suggestion
The imports and query declarations are changed to imports for the JSON that correspond to the query result that Gatsby found when it ran the query. Consider the following component with with a static query written using the `useStaticQuery` hook:
```

laurieontech commented:
I love having an explanation for this! These are really good examples and clarifications. Think this post will have a big impact on removing some of the gatsby magic!

gillkyle commented on docs/docs/static-vs-normal-queries.md:
````diff
@@ -45,18 +31,97 @@ The redux `staticQueryComponents` is a map from component jsonName to StaticQuer
+}
+```
+
+Will have the query string removed and replaced with an import for the JSON file created and named with its specific hash. The Redux stores tracking the queries link the correct data to the page they correspond to.
````
```suggestion
This component will have the query string removed and replaced with an import for the JSON file created and named with its specific hash. The Redux stores tracking the queries link the correct data to the page they correspond to.
```

muescha commented:
Note: all links in docs for query-runner has changed. I created a follow up issue #19557 to decouple it from this PR

laurieontech commented on docs/docs/static-vs-normal-queries.md:
```diff
@@ -2,35 +2,21 @@
-> - describe how queries are stripped and JSON imports are rewritten
->
-> You can help by making a PR to [update this documentation](https://github.com/gatsbyjs/gatsby/issues/14228).
+Gatsby handles 3 varieties of GraphQL queries: Page queries (sometimes for simplicity referred to as "normal queries"), static queries using the `<StaticQuery />` component, and static queries using the `useStaticQuery` hook.
```
```suggestion
Gatsby handles three varieties of GraphQL queries: Page queries (sometimes for simplicity referred to as "normal queries"), static queries using the `<StaticQuery />` component, and static queries using the `useStaticQuery` hook.
```

laurieontech commented:
Looks great @gillkyle 🎉
